### ISSUE TITLE
Will this be a NextPush replacement?

### ISSUE THREAD

**Issue by woiwoiwoiwoi** [bitfireAT/nc_ext_dav_push#33]
Let me first say that I'm a big fan of your work. DAVx5 is amazing, thank you for creating it!

I'm looking to move away from Google services as much as possible, and since a first few apps I use have started supporting UnifiedPush, I've started looking into enabling that on my Nextcloud instance. From what I understand, once this extension is more stable, I should be able to use it instead of NextPush or a selfhosted ntfy server and other UnifiedPush-compatible Android apps should be able to use it, correct?

I'm in no immediate rush to set this up. In fact, I prefer to wait for a solution that solves the most problems for me. And I believe that this extension would work for DAVx5, the Nextcloud Android app, the Nextcloud Desktop app, as well as any UnifiedPush-enabled Android apps?

**Comment by ArnyminerZ** [comment]
Hi, thanks to you for using DAVx⁵. We are not aiming to provide any Push server (such as ntfy or NextPush) at least right now. Our work focuses on creating an standard for webdav (https://github.com/bitfireAT/webdav-push) that would allow everyone to move away from proprietary solutions for calendar, contacts and files.

This repo is just an example implementation of what's possible, with Nextcloud, which is what we mostly use, but it can be implemented anywhere, as long the server provider implements the standard.

But yeah, in short, hopefully, the final idea is that this is an standard, and works with everything, not just Nextcloud, or DAVx⁵, but with any WebDAV server.

**Comment by woiwoiwoiwoi** [comment]
Thanks for the quick response!

So the plan is that once your standard is standardised (parden the pun), NextPush or ntfy include it in their servers, right? I would then selfhost one of those two and my DAVx5 (along with all other UnifiedPush compatible apps) would also be able to use them?

**Comment by ArnyminerZ** [comment]
Technically UnifiedPush is a protocol in itself (based on other protocols and standards), you can find how it works [on their website](unifiedpush.org), but the interesting thing is that it allows you to have your own server. So let's say you have a Nextcloud Server with CalDAV support, and you want to add support for Push. Right now you have to install this extension, and choose a UnifiedPush distributor (there are a lot, [see some examples](https://unifiedpush.org/users/distributors/)). It doesn't matter which one you use, it doesn't have to be NextPush in order to work with Nextcloud.

This extension, communicates with the client (DAVx⁵ for example), and negotiates the distributor to use, which then is used for sending push messages.

Ideally, and this in theory will happen at some point in the future, this repository will no longer be an extension, but it will be built into Nextcloud (as well as other servers hopefully), and you just would have to select a distributor from your client. Then, the client tells the server "hey, I support push", and if the server also supports push (which would be the case), the communication would be established.

On the client side, this can even be embedded into the app, so you don't even have to install a distributor in your phone, the built into the app would be used. It can be [FCM](https://firebase.google.com/docs/cloud-messaging) also, for example.

I hope this makes everything much clearer. We gave a [lightning talk at NextcloudConf this year](https://youtu.be/6fX_q8UGIYU?si=UiuGcBf57kgkcHo2) if you want to have a clearer view.

**Comment by woiwoiwoiwoi** [comment]
Thanks again for the elaborate answer. I actually had watched your lightning talk before, and that's how I got excited and ended up here in this repo!

Do I understand it correctly that these are the current and future options:

1. install this extension and NextPush (or ntfy) on Android -> get real-time DAVx5 updates  (calendar, contacts, todos), but no other UnifiedPush compatible Android apps can use it
2. install the UnifiedPush Nextcloud server app (or ntfy server), NextPush (or ntfy) on Android -> get real-time notifications in UnifiedPush compatible Android apps, but not in DAVx5 (calendar, contacts, todos)
3. install this extension, the UnifiedPush Nextcloud server app (or ntfy server), NextPush (or ntfy) on Android -> get real-time notifications for DAVx5 and all other UnifiedPush compatible Android apps
4. a future is possible where only a regular Nextcloud server (without the UnifiedPush app) and DAVx5 are required for real-time calendar/contacts/todo updates - but if one wants other UnifiedPush compatible Android apps to get real-time notifications, ntfy server and ntfy Android would still be required, right?

**Comment by ArnyminerZ** [comment]
> but no other UnifiedPush compatible Android apps can use it

No, that's the good thing, other UnifiedPush-compatible apps can use it. The thing about NextPush is that it's easy to self-host with Nextcloud, not that it's only compatible with Nextcloud.

> and NextPush (or ntfy) on Android

To use NextPush I believe you must have a Nextcloud server with NextPush installed. Just to have into account.

> but not in DAVx5

Yes, if you don't install this extension on Nextcloud, DAVx⁵ won't have push notifications.

> 3. install this extension, the UnifiedPush Nextcloud server app (or ntfy server), NextPush (or ntfy) on Android -> get real-time notifications for DAVx5 and all other UnifiedPush compatible Android apps

I would join this with the first option. There's no NextPush without installing it in Nextcloud.

> 4. a future is possible where only a regular Nextcloud server (without the UnifiedPush app) and DAVx5 are required for real-time calendar/contacts/todo updates - but if one wants other UnifiedPush compatible Android apps to get real-time notifications, ntfy server and ntfy Android would still be required, right?

There's a _near_ future in which only a UnifiedPush distributor is needed. I mean that you won't have to install this extension in Nextcloud, it would be built-in, in fact, technically this repo will not exist at that point, or if it exists, it will be as "sample server implementation".

---

NextPush/nfty are simply UnifiedPush distributors, they are basically the same thing, though ntfy goes through a shared server, and with NextPush you can self-host it. They don't have anything to do with DAVx⁵ and/or Nextcloud. The only thing they have in relation is that NextPush is a Nextcloud app, but it doesn't give any extra functionality to Nextcloud, it's just an easy installer.

Then, you have this extension, which does work in consonance with Nextcloud, because it extends the functionalities given by Nextcloud (CalDAV/CardDAV/WebDAV), and DAVx⁵ (because it's the only client compatible with WebDAV-Push _right now_).

But again, WebDAV-Push should not be related in any way to Nextcloud (just like WebDAV is not related in any way with Nextcloud). It's a protocol used by Nextcloud, but it isn't related to it.

**Comment by devvv4ever** [comment]
To give another example: I'm using the UnifiedPush implementation from Conversations, my XMPP messenger on Android, since I don't have any other ntfy or Nextcloud app installed on my device. Conversations also provides a UnifiedPush implementation, and I've selected this distributor in my DAVx5 to be used. So the push of an updated collection on Nextcloud is going through UP by Conversations which then notifies DAVx5 that there are changes - and then the sync is enqueued by DAVx5.

**Comment by woiwoiwoiwoi** [comment]
Thanks for all your responses. I really got into this now. Have been reading up on it as much as I could and will test things myself in a bit. One or two quick practical questions:
- can I revert to my previous setup using the standard 15 minute polling interval? can I simply uninstall this extension without complications?
- maybe you have more experience with ntfy and NextPush: I would prefer using NextPush (less configuring required), but I read that ntfy also offers things like notifications from cron jobs and various other interesting ways to use it to deliver notifications. Are these things also possible with NextPush or are they ntfy features (couldn't find anything on that)?

**Comment by ArnyminerZ** [comment]
> can I revert to my previous setup using the standard 15 minute polling interval? can I simply uninstall this extension without complications?

Regarding Nextcloud I'm not sure about that, I think there were some issues regarding database cleanup, but @JonathanTreffler can answer better. Regarding DAVx⁵, yes, you can uninstall it, or disable the feature completely without any issues (that we know of 😅).

Feel free to report any issues that arise.

> maybe you have more experience with ntfy and NextPush: I would prefer using NextPush (less configuring required), but I read that ntfy also offers things like notifications from cron jobs and various other interesting ways to use it to deliver notifications. Are these things also possible with NextPush or are they ntfy features (couldn't find anything on that)?

I don't know to be honest.

**Comment by JonathanTreffler** [comment]
> can I revert to my previous setup using the standard 15 minute polling interval? can I simply uninstall this extension without complications?

Yes. The dav push app can just be disabled or deleted and the nextcloud dav server will no longer advertise the push feature to clients like davx5. Uninstalled nextcloud apps never delete their data in the database, so that if you re-install it all your data is still there, but for his app this is irrelevant, because the subscriptions will have probably expired by then anyways.
